qp/utils.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import sys
+
+def safelog(arr, threshold=sys.float_info.epsilon):
+    """
+    Takes log of array with zeroes.
+
+    Parameters
+    ----------
+    arr: ndarray
+        Values to be logged
+    threshold: float
+        Small, positive value to replace zeros and negative numbers
+
+    Returns
+    -------
+    logged: ndarray
+        Logged values, with approximation in place of zeros and negative numbers
+    """
+    shape = np.shape(arr)
+    flat = arr.flatten()
+    logged = np.log(np.array([max(a,threshold) for a in flat])).reshape(shape)
+    return logged
+
+def calckl(qn, pn):
+    """
+    Calculates Kullback-Leibler Divergence
+
+    Parameters
+    ----------
+    qn: ndarray
```

Comment:
Well - revisiting the (I think pretty good) wikipedia page, I read that:

> KL divergence is a measure of the difference between two probability distributions P and Q. It is not symmetric in P and Q. In applications, P typically represents ... a precisely calculated theoretical distribution, while Q typically represents ... [an] approximation of P.
>
> Specifically, the Kullback–Leibler divergence from Q to P, denoted DKL(P‖Q), is a measure of the information gained when one revises one's beliefs from ... Q to ... P. In other words, it is the amount of information lost when Q is used to approximate P.

I have carefully excised some confusing words while preserving the piece that is relevant to us here. P could be a histogram representation and Q a quantile approximation, and the KL divergence would be the amount of information lost when quantiles are used to approximate the PDF, relative to the histogram approximation. I agree that the asymmetry is a bit tricky. But in the case where P is a true distribution, the information loss statement makes more sense. So, when testing, you could compare each of your approximations with the truth via the KLD, and then compare the information lost in each approximation to make a decision about which is best.